gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc

```diff
@@ -0,0 +1,143 @@
+  {
+    std::ifstream template_file(template_file_path);
+    std::stringstream buffer;
+    buffer << template_file.rdbuf();
+    template_content = buffer.str();
+  }
+
+  const std::unique_ptr<gematria::LlvmArchitectureSupport> llvm_support =
+      gematria::LlvmArchitectureSupport::X86_64();
+  gematria::X86Canonicalizer canonicalizer(&llvm_support->target_machine());
+  gematria::BHiveImporter bhive_importer(&canonicalizer);
+
+  std::ifstream bhive_csv_file(bhive_filename);
+  for (std::string line; std::getline(bhive_csv_file, line);) {
+    auto comma_index = line.find(',');
+    if (comma_index == std::string::npos) {
+      std::cerr << "Invalid CSV file: no comma in line '" << line << "'\n";
+      return 2;
+    }
+
+    std::string_view hex = std::string_view(line).substr(0, comma_index);
+    auto bytes_or = gematria::ParseHexString(hex);
+    if (!bytes_or.has_value()) {
+      std::cerr << "could not parse: " << hex << "\n";
+      return 3;
+    }
+
+    // for each line, find the accessed addresses & disassemble instructions
+    const auto& bytes = bytes_or.value();
+    auto addrs_or = gematria::FindAccessedAddrs(bytes);
+    auto proto = bhive_importer.BasicBlockProtoFromMachineCode(bytes);
+    if (addrs_or.ok() && proto.ok()) {
+      // TODO: attach those addresses to input that can be executed by
+      // llvm-exegesis Create output file path
+      std::string output_file_path =
+          output_dir + "/" + std::to_string(file_counter) + ".test";
+
+      // Open output file for writing
+      std::ofstream output_file(output_file_path);
+      if (!output_file.is_open()) {
+        std::cerr << "Failed to open output file: " << output_file_path << "\n";
+        return 4;
+      }
+
+      // Write the template content into the output file
+      output_file << template_content;
+
+      // Append memory annotations
+      auto addrs = addrs_or.value();
+      for (const auto& addr : addrs.accessed_blocks) {
+        std::string mem_name =
+            LLVM_EXEGESIS_MEM_NAME_PREFIX + std::to_string(mem_counter++);
+        output_file << LLVM_EXEGESIS_MEM_DEF_PREFIX << mem_name << " "
+                    << addrs.block_size << " " << INITIAL_MEM_VALUE << "\n";
+        output_file << LLVM_EXEGESIS_MEM_MAP_PREFIX << mem_name << " "
+                    << std::dec << addr << "\n";
+      }
+
+      // Append disassembled instructions
+      for (const auto& instr : proto->machine_instructions()) {
+        output_file << instr.assembly() << "\n";
+      }
+
+      file_counter++;
+    } else if (proto.ok()) {
+      std::cerr << "Failed to find addresses for block '" << hex
+                << "': " << addrs_or.status() << "\n";
+      std::cerr << "Block disassembly:\n";
+      for (const auto& instr : proto->machine_instructions()) {
+        std::cerr << "\t" << instr.assembly() << "\n";
+      }
+    } else {
+      std::cerr << "Failed to disassemble block '" << hex << "\n";
+    }
```

Comment:
I'd prefer to put this to the beginning of the function, with a `continue` if there is a non-critical failure. Also, we should print the status to make debugging easier.

```
if (!proto.ok()) {
  std::cerr << "Failed to disassemble block '" << hex << "': " << proto.status() << "\n";
  continue;
}
// ...
if (!addrs.ok()) { 
  std::cerr << "Failed to determine used addresses for block '" << hex << "': "
<< addrs.status() << " \n";
  // ...print disassembly...
  continue;
}
```